app/services/custom_fields/hierarchy/hierarchical_item_formatter.rb

```diff
@@ -0,0 +1,105 @@
+# frozen_string_literal: true
+
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module CustomFields
+  module Hierarchy
+    class HierarchicalItemFormatter
+      include NumberFormatHelper
+
+      # @param [Boolean] path Show the full item ancestor path if true
+      # @param [Boolean] label Show the item label if true
+      # @param [Boolean] suffix Show the item short or weight if true
+      # @param [Boolean] suffix_parentheses Wrap suffix in parentheses.
+      # @param [Integer] number_length_limit If formatting numbers, restrict the whole length of the number to the
+      # given limit before falling back to scientific notation.
+      # @param [Integer] number_integer_digit_limit If formatting numbers, restrict the integer part of the number to
+      # the given limit before falling back to scientific notation.
+      # @param [Integer] number_precision If formatting numbers, round to the given precision.
+      def initialize(path: false,
+                     label: true,
+                     suffix: true,
+                     suffix_parentheses: true,
+                     number_length_limit: 9,
+                     number_integer_digit_limit: 8,
+                     number_precision: 4)
+        @path = path
+        @label = label
+        @suffix = suffix
+        @suffix_parentheses = suffix_parentheses
+        @number_length_limit = number_length_limit
+        @number_integer_digit_limit = number_integer_digit_limit
+        @number_precision = number_precision
+      end
+
+      def format(item:)
+        parts = []
+
+        if @path && @label
```

Comment:
~~:yellow_circle: Why is this constructed as a three-case if-elsif-elsif? As far as I can tell, we only use two inputs that may or may not be concatenated.~~

I'd have expected something like

```ruby
parts << ancestors(item) if @path
parts << item.label if @label
```

edit: Okay I see it now. `parts` is for all visible parts of the formatted thing, not just the path components and concatenating the path with `/` already happens in `branch` and `ancestors`... I guess it makes sense, but it surprised me.